[로그인] 소개 입력란 페이지 UI

WhiteHyun opened the pull request:
## 📌 PR 요약

🌱 작업한 내용
- UI를 구현했습니다.

🌱 PR 포인트
- UITextView는 `placeholder`를 지원하지 않아 `textColor`와 `text`를 이용하여 눈속임으로 `placeholder`처럼 보이도록 구현했습니다.
- UITextView에 `text`가 입력될 때마다 다음 줄로 넘어갈 시 `text`가 차지하는 사이즈만큼 textView의 `height`를 계산하여 view를 update해주어야했습니다. 다행히 `sizeThatFits(_:)`메서드가 존재하였고, 이를 이용하여 textView의 `dynamic height`를 설정할 수 있었습니다.

## 📸 스크린샷
|스크린샷|
|:--:|
|![Screen Recording 2023-01-04 at 4 38 35 PM](https://user-images.githubusercontent.com/57972338/210507362-8e1dd4f9-e9c1-4d6d-af08-1233d28d9e78.gif)|


## 📮 관련 이슈
- #16

WhiteHyun committed:
[ADD] IntroductionVC

WhiteHyun committed:
[ADD] StackView

WhiteHyun committed:
[ADD] Introduction label

WhiteHyun committed:
[ADD] text view to write user's introduction

WhiteHyun committed:
[ADD] Placeholder string property

WhiteHyun committed:
[FEAT] Placeholder in TextView

TextView에서는 placeholder를 지원하지 않기 때문에
text를 이용한 눈속임으로 Placeholder처럼 보이도록 구현해두었습니다.
이를 위해 delegate를 채택하여 알맞은 조건에 placeholder처럼 보이도록 구현해야했습니다.

WhiteHyun committed:
[ADD] Corner radius in text view

WhiteHyun committed:
[ADD] Method resigning first responder

WhiteHyun committed:
[ADD] overline label

입력한 글자수와 제한하는 글자 수를 표기하기위한 label입니다.
text에 서로 다른 색상이 들어가있어 NSAttributeString을 이용했습니다.

WhiteHyun committed:
[ADD] Constant property called `totalCharacterLimit`

총 글자 제한 수를 나타내는 프로퍼티입니다.

WhiteHyun committed:
[FEAT] Dynamic label text based on TextView text count

text가 변동될 때마다 글자 수만큼 overline label의 text도 바뀌도록 구현했습니다.

WhiteHyun committed:
[FEAT] TextView height that dynamically updates as text is entered

WhiteHyun committed:
[FIX] TextView not being cleared when reaching 150 characters

WhiteHyun committed:
[CHORE] Stack view's spacing

WhiteHyun committed:
[DEL] bind()

dlrjswns commented on PLUB/Sources/Views/Login/Introduction/IntroductionViewController.swift:
```diff
@@ -0,0 +1,150 @@
+extension IntroductionViewController: UITextViewDelegate {
+  func textViewDidEndEditing(_ textView: UITextView) {
+    // text가 비어있으면서 글자색이 검정인 경우 -> 사용자가 입력 후 다 지운 경우
+    if textView.text.isEmpty && textView.textColor == .black {
```
이 부분에 trimming으로 white space와 line까지 없앤 text값의 isEmpty값을 확인해주면 더 좋을꺼같아요 !!😁😁

WhiteHyun replied:
시간내어 리뷰 달아주셔서 감사합니다. ☺️ 제가 이 생각을 못했네요!! 조언 감사드립니다. 수정하고 다시 리뷰 요청 남기겠습니다 :)

WhiteHyun replied:
말씀하신 부분을 구현하면서 오류 케이스들이 추가로 발견되었고, 수정하는 과정에서 커밋이 몇 개 추가되었습니다.

1. whitespace + newline만 존재할 시 isEmpty로 처리함
   - 입력된 글자 수를 보여주는 overline label이 업데이트되지 않는 현상 수정
   - textView의 높이가 다시 원상태로 복귀하지 않는 현상 수정

|전|후|
|:-:|:-:|
|![Screen Recording 2023-01-06 at 5 02 30 PM](https://user-images.githubusercontent.com/57972338/210957196-ce826d66-110d-417f-a751-aad0cece76f1.gif)|![Screen Recording 2023-01-06 at 4 26 23 PM](https://user-images.githubusercontent.com/57972338/210957118-ff64e0da-a9e9-4669-9adf-ebc9f17a9927.gif)|

dlrjswns replied:
승현님 혹시 위 1번 사항에 textView초기화하는 부분 수정코드가 어디에 있을까요 ?? 👀👀

WhiteHyun replied:
이쪽에서 if문 분기에 들어가는 구간이 `1번 사항`입니다.
https://github.com/PLUB2022/PLUB-iOS/blob/f686fa8e66fb25d29c0b74c66dc8d63d32d6472a/PLUB/Sources/Views/Login/Introduction/IntroductionViewController.swift#L120-L128

dlrjswns commented on PLUB/Sources/Views/Login/Introduction/IntroductionViewController.swift:
```diff
@@ -0,0 +1,150 @@
+  func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+    // if the user presses the delete key, the length of the range is 1.
+    if range.length == 1 { return true }
+    return textView.text.count < totalCharacterLimit ? true : false
```
승현님은 진짜 디자이너분들이 요청한 이런 세세한 부분까지 안놓치시는군요 !! 완전 대단하십니다 👀👀

soobin-k commented on PLUB/Sources/Views/Login/Introduction/IntroductionViewController.swift:
```diff
@@ -7,24 +7,139 @@
 
+// MARK: - UITextViewDelegate
+
+extension IntroductionViewController: UITextViewDelegate {
```
UITextViewDelegate 대신 RxSwift의 textview.rx.didBeginEditing, .didEndEditing, .text 사용해보시면 더 좋을 것 같습니다!
코드가 더 간결해질 것 같아서요😊

WhiteHyun replied:
리뷰 감사합니다 :) 얼른 구현하러 갈게용

WhiteHyun replied:
마냥 RxSwift가 좋지만은 않군요 😭
두 가지 옵션에 대해서만 코드를 작성해보았는데 저는 전자가 더 읽기 쉬운 것 같아요 :')

- <details><summary>전</summary>

  ```swift
  extension IntroductionViewController: UITextViewDelegate {
    func textViewDidEndEditing(_ textView: UITextView) {
      // text가 비어있으면서 글자색이 검정인 경우 -> 사용자가 입력 후 다 지운 경우
      if textView.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && textView.textColor == .black {
        textView.text = placeHolder
        textView.textColor = .deepGray
      }
    }
    
    func textViewDidBeginEditing(_ textView: UITextView) {
      // text가 지정된 placeholder면서 글자색이 짙은회색인 경우
      // 즉 placeholder을 띄운 상태인데 사용자가 수정하려고 textview를 누른 경우
      if textView.text == placeHolder && textView.textColor == .deepGray {
        textView.text = ""
        textView.textColor = .black
      }
    }
  }
  ```

   </details>


- <details><summary>후</summary>

  ```swift
  override func bind() {
    super.bind()
    
    self.introductionTextView.rx.didEndEditing
      .asDriver()
      .drive(onNext: { [weak self] in
        guard let self = self else { return }
        // text가 비어있으면서 글자색이 검정인 경우 -> 사용자가 입력 후 다 지운 경우
        if self.introductionTextView.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && self.introductionTextView.textColor == .black {
          self.introductionTextView.text = self.placeHolder
          self.introductionTextView.textColor = .deepGray
        }
      })
      .disposed(by: disposeBag)
    
    self.introductionTextView.rx.didBeginEditing
      .asDriver()
      .drive(onNext: { [weak self] in
        guard let self = self else { return }
        // text가 지정된 placeholder면서 글자색이 짙은회색인 경우
        // 즉 placeholder을 띄운 상태인데 사용자가 수정하려고 textview를 누른 경우
        if self.introductionTextView.text == self.placeHolder && self.introductionTextView.textColor == .deepGray {
          self.introductionTextView.text = ""
          self.introductionTextView.textColor = .black
        }
      })
      .disposed(by: disposeBag)
  }
  ```
 </details>

추가로 [RxSwift/UITextView+Rx.swift at main · ReactiveX/RxSwift](https://github.com/ReactiveX/RxSwift/blob/main/RxCocoa/iOS/UITextView%2BRx.swift)에서 delegate에 존재하는 4가지 옵션만 API를 제공해주는데, 

`textViewDidChange(_:)`와 `textView(_:shouldChangeTextIn:replacementText:)`와 같은 text값이 바뀌는 흐름에 관한 API가 없더라구요. 저는 각 메서드 역할에 맞게 구현해둔 상태라 RxSwift+UITextView extension으로 ControlEvent를 추가하거나 해야할 듯 싶습니다.
그러면 결국엔 delegate보다 코드가 길어지므로.. rx와 delegate(또는 dataSource)는 적절한 시점에 사용해야할 듯 싶습니다(🤔)

soobin-k replied:
rx를 사용할 때 더 불편한 점이 있는지 몰랐네요
승현님이 보여주신 예제 코드를 보니 delegate를 활용하는게 더 좋아보입니다.
깊이 고민해주셔서 감사합니다!👍

soobin-k commented on PLUB/Sources/Views/Login/Introduction/IntroductionViewController.swift:
```diff
@@ -7,24 +7,139 @@
+  func textViewDidBeginEditing(_ textView: UITextView) {
+    // text가 지정된 placeholder면서 글자색이 짙은회색인 경우
+    // 즉 placeholder을 띄운 상태인데 사용자가 수정하려고 textview를 누른 경우
+    if textView.text == placeHolder && textView.textColor == .deepGray {
```
if의 textView.text == placeHolder 이 조건만으로도 placeholder인지 아닌지 판단이 가능해보입니다!

WhiteHyun replied:
사용자가 "소개하는 내용을 입력해주세요"를 입력하고 First responder를 resign하게 되면 다시 수정하려 할 때 text가 제거됩니다.
물론 사용자가 저걸 그대로 입력하지는 않겠지만(하하), 이를 염두하여 추가의 조건을 추가했습니다.

soobin-k replied:
제가 놓친 부분이 있었네요! 자세한 설명 감사합니다!

soobin-k commented on PLUB/Sources/Views/Login/Introduction/IntroductionViewController.swift:
```diff
@@ -7,24 +7,139 @@
+    let estimatedSize = textView.sizeThatFits(size)
+
+    // update height constraints
+    textView.snp.updateConstraints { make in
```
make in 을 생략하고 다음줄의 make를 $0로 바꾸면 코드가 더 보기 좋아질 것 같습니다!

WhiteHyun replied:
저희가 기존 코드를 작성하기 전에 해당 스타일에 대한 의견이 있었는데, make를 사용하는 게 코드가 영문으로 읽혀서 사용하는 것으로 코드 스타일을 지정했었습니다. 아니면, 금주에 다시 한 번 모여서 코드 스타일에 대해 서로 의견을 나눠볼까요?

soobin-k replied:
넵 좋습니다! 이번주 팀회의 시간에 같이 논의해봐요!

WhiteHyun committed:
[FIX] Check for empty textView after trimming whitespace and newlines

WhiteHyun committed:
[ADD] updateWrittenCharactersLabel function

작성된 글자 수를 업데이트할 수 있도록 하는 메서드를 구현했습니다.
overlineLabel 내에서, delegate 내에서 비슷하게 코드가 반복되고 있고,
추가로 더 작성될 곳도 존재하기에 함수로 구조를 만들어두었습니다.

WhiteHyun committed:
[FIX] Update overline label when user clears textView input

WhiteHyun committed:
[CHORE] Overline label's point color when user deleted all texts

텍스트 값이 0일 경우에는 무조건 mediumGray 색으로 보이게 수정했습니다.
그 외에는 이제 검은색으로 보여집니다.

WhiteHyun committed:
[FEAT] `updateTextViewHeightAutomatically` function

기존 textViewDidChange(_:) 내부에 구현된 코드를 함수로 적용시켰습니다.

WhiteHyun committed:
[FIX] Issue where textView height is not reset when text is empty

WhiteHyun committed:
Merge branch 'develop' into feat/16-Register/Introduction
